python/ray/ml/utils/remote_storage.py

```diff
@@ -0,0 +1,165 @@
+import urllib.parse
+from typing import Optional, Tuple
+
+try:
+    import fsspec
+except ImportError:
+    fsspec = None
+
+try:
+    import pyarrow
+    import pyarrow.fs
+except (ImportError, ModuleNotFoundError):
+    pyarrow = None
+
+from ray import logger
+
+# We keep these constants for legacy compatibility with Tune's sync client
+# After Tune fully moved to using pyarrow.fs we can remove these.
+S3_PREFIX = "s3://"
+GS_PREFIX = "gs://"
+HDFS_PREFIX = "hdfs://"
+ALLOWED_REMOTE_PREFIXES = (S3_PREFIX, GS_PREFIX, HDFS_PREFIX)
+
+
+def _assert_pyarrow_installed():
+    if pyarrow is None:
+        raise RuntimeError(
+            "Uploading, downloading, and deleting from cloud storage "
+            "requires pyarrow to be installed. Install with: "
+            "`pip install pyarrow`. Subsequent calls to cloud operations "
+            "will be ignored."
+        )
+
+
+def fs_hint(uri: str) -> str:
+    """Return a hint how to install required filesystem package"""
+    if pyarrow is None:
+        return "Please make sure PyArrow is installed: `pip install pyarrow`."
+    if fsspec is None:
+        return "Try installing fsspec: `pip install fsspec`."
+
+    from fsspec.registry import known_implementations
+
+    protocol = urllib.parse.urlparse(uri).scheme
+
+    if protocol in known_implementations:
+        return known_implementations[protocol]["err"]
+
+    return "Make sure to install and register your fsspec-compatible filesystem."
```

Comment:
In this case actually all returns are error conditions - the file system URI (e.g. `foo://bar`) could not be associated to a installed filesystem, so we raise an error how to fix this.

We thus follow a progression here - pyarrow is the base and required for everything, so that is checked first. If pyarrow is installed but the URI is not found (pyarrow only supports local, s3, hdfs), the filesystem can be provided with fsspec. Thus the user should install fsspec next if it's not already there.

fsspec provides a dict `from fsspec.registry import known_implementations` which contains all fsspec-compatible filesystems that are "known" to the fsspec maintainers. Here is the current definition: https://github.com/fsspec/filesystem_spec/blob/master/fsspec/registry.py#L87

For instance, for `dropbox://something`, if it's not available, the user would get the message
```
            'DropboxFileSystem requires "dropboxdrivefs",'
            '"requests" and "dropbox" to be installed'
```

Only as a last resort we would end with the generic "Make sure to install and register your fsspec-compatible filesystem.", which should usually only come up when a user wants to use a custom URI and filesystem implementation.

IMO this order makes sense and we should keep it that way